### ISSUE TITLE
[Bug]: `allowed_token_ids` + `bad_words` can emit tokens outside the allowlist

### ISSUE THREAD

**Issue by Ningke-Li** [vllm-project/vllm#57986]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Collecting environment information...
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version                : Could not collect
CMake version                : version 3.22.1
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.13.0+cu130
Is debug build               : False
CUDA used to build PyTorch   : 13.0
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.14 | packaged by Anaconda, Inc. | (main, Aug 27 2026, 14:46:43) [GCC 14.3.0] (64-bit runtime)
Python platform              : Linux-5.15.0-122-generic-x86_64-with-glibc2.35
    
==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   : 
GPU models and configuration : 
GPU 0: NVIDIA L40S
GPU 1: NVIDIA L40S
GPU 2: NVIDIA L40S
GPU 3: NVIDIA L40S

Nvidia driver version        : 595.71.05
cuDNN version                : Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : False

==============================
          CPU Info
==============================
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        52 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               128
On-line CPU(s) list:                  0-127
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 9554 64-Core Processor
CPU family:                           25
Model:                                17
Thread(s) per core:                   2
Core(s) per socket:                   64
Socket(s):                            1
Stepping:                             1
Frequency boost:                      enabled
CPU max MHz:                          3762.9880
CPU min MHz:                          1500.0000
BogoMIPS:                             6190.64
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d
Virtualization:                       AMD-V
L1d cache:                            2 MiB (64 instances)
L1i cache:                            2 MiB (64 instances)
L2 cache:                             64 MiB (64 instances)
L3 cache:                             256 MiB (8 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-127
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.16.post3
[pip3] nccl4py==0.5.0
[pip3] numpy==2.3.5
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cccl==13.3.3.4.1
[pip3] nvidia-cuda-crt==13.3.73
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvcc==13.3.73
[pip3] nvidia-cuda-nvdisasm==13.3.73
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cudnn-frontend==1.28.0
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufile==1.15.1.6
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-cutlass-dsl==4.6.2
[pip3] nvidia-cutlass-dsl-libs-base==4.6.2
[pip3] nvidia-cutlass-dsl-libs-core==4.6.2
[pip3] nvidia-cutlass-dsl-libs-cu12==4.6.2
[pip3] nvidia-cutlass-dsl-libs-cu13==4.6.2
[pip3] nvidia-ml-py==13.610.43
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.3.33
[pip3] nvidia-nvshmem-cu13==3.4.5
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvvm==13.3.73
[pip3] pyzmq==27.2.0
[pip3] tokenspeed-triton==3.8.10.post20260906
[pip3] torch==2.13.0
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torchaudio==2.11.0
[pip3] torchcodec==0.16.0
[pip3] torchvision==0.28.0
[pip3] transformers==5.16.1
[pip3] triton==3.7.1
[conda] numpy                        2.4.3            pypi_0              pypi

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.29.0
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled; XPU: Disabled
GPU Topology:
  	GPU0	GPU1	GPU2	GPU3	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	PHB	NODE	NODE	0-127	0		N/A
GPU1	PHB	 X 	NODE	NODE	0-127	0		N/A
GPU2	NODE	NODE	 X 	PHB	0-127	0		N/A
GPU3	NODE	NODE	PHB	 X 	0-127	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
NO_COLOR=1
CUDA_HOME=/usr/local/cuda
CUDA_HOME=/usr/local/cuda
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_xxx

```

</details>

## Description

Combining two supported sampling constraints can leave no valid next token and silently break the allowlist.

With Qwen3-0.6B, the text `" a"` is token 264. This request allows only token 264 and also bans `" a"`:

```json
{
  "model": "Qwen/Qwen3-0.6B",
  "prompt": "Return a short answer.",
  "temperature": 0,
  "max_tokens": 4,
  "return_token_ids": true,
  "allowed_token_ids": [264],
  "bad_words": [" a"]
}
```

vLLM accepts the request and returns HTTP 200, but generates token 0 four times. Token 0 is not in `allowed_token_ids`.

This is not an empty `allowed_token_ids=[]` request. Both fields are non-empty, each constraint works on its own, and vLLM does not report the combination as unsupported.

Applications use `allowed_token_ids` to force outputs into a fixed vocabulary, for example label IDs, routing choices, or constrained decoding. This bug can return an arbitrary out-of-domain token in a successful response, so clients that trust the allowlist can accept an invalid result without seeing an error.

## Reproduction

Start the server:

```bash
CUDA_VISIBLE_DEVICES=0 \
VLLM_USE_FLASHINFER_SAMPLER=0 \
vllm serve Qwen/Qwen3-0.6B \
  --served-model-name Qwen/Qwen3-0.6B \
  --host 127.0.0.1 \
  --port 8032 \
  --gpu-memory-utilization 0.25 \
  --max-model-len 256 \
  --max-num-seqs 8 \
  --enforce-eager \
  2>&1 | tee server.log
```

Use the script [repro.py](https://github.com/user-attachments/files/32479783/repro_empty_sampler.py), run:

```bash
python repro.py \
  --base-url http://127.0.0.1:8032 \
  2>&1 | tee reproducer.log
```

The script obtains the token IDs from the model tokenizer and sends four requests: each constraint alone, the empty intersection, and a non-empty intersection control. It exits with status 1 when the out-of-allowlist result is reproduced.

## Actual result

All four requests return HTTP 200:

| Case | Constraints | Returned token IDs | Text |
| --- | --- | --- | --- |
| Allowlist only | allow `[264]` | `[264, 264, 264, 264]` | `" a a a a"` |
| Bad word only | ban token 264 | `[4710, 785, 2701, 374]` | `" \n\nThe following is"` |
| Empty intersection | allow `[264]`, ban token 264 | `[0, 0, 0, 0]` | `"!!!!"` |
| Non-empty control | allow `[264, 279]`, ban token 264 | `[279, 279, 279, 279]` | `" the the the the"` |

The controls show that the individual processors work. Only the contradictory combination returns tokens outside the allowlist. The complete client and server output is in [server.log](https://github.com/user-attachments/files/32479792/server_empty_sampler.log).

## Expected result

If the active constraints leave no valid next token, vLLM should reject the request or stop it with a clear no-valid-token error. A successful response must never contain a token outside `allowed_token_ids`.

## Root cause

The V2 sampler applies the two masks sequentially:

1. The [`allowed_token_ids` kernel](https://github.com/vllm-project/vllm/blob/v0.29.0/vllm/v1/worker/gpu/sample/logit_bias.py#L193-L226) sets every logit to `-inf`, then restores token 264.
2. [`bad_words` runs afterward](https://github.com/vllm-project/vllm/blob/v0.29.0/vllm/v1/worker/gpu/sample/sampler.py#L224-L245) and [sets token 264 back to `-inf`](https://github.com/vllm-project/vllm/blob/v0.29.0/vllm/v1/worker/gpu/sample/bad_words.py#L148-L166).
3. There is no check for an all-`-inf` row. At temperature 0, the [Gumbel sampler performs a plain argmax](https://github.com/vllm-project/vllm/blob/v0.29.0/vllm/v1/worker/gpu/sample/gumbel.py#L102-L130), and the all-tied row falls through to token 0.

Request validation rejects a literal empty allowlist and out-of-vocabulary IDs, but [does not validate the intersection with `bad_words`](https://github.com/vllm-project/vllm/blob/v0.29.0/vllm/sampling_params.py#L981-L1010).

## Related issues

[#53665](https://github.com/vllm-project/vllm/issues/53665) reports a different way to create an all-masked row: a terminal structured-output grammar combined with `min_tokens`.

[#54218](https://github.com/vllm-project/vllm/pull/54218) fixes that specific `min_tokens + structured output` case. Its description mentions that semantic masks such as `allowed_token_ids` and `bad_words` can cause the same broader all-`-inf` problem, but explicitly leaves the general fix for future work. The implementation only enables stop-token restoration when both `min_tokens` and structured output are active; it does not detect the intersection reproduced here. This report was reproduced on vLLM 0.29.0, which already contains that PR. The same processor ordering and lack of cross-constraint validation are still present on the current `main` branch at the time of writing.

No dedicated issue for the `allowed_token_ids + bad_words` intersection was found in searches of existing vLLM issues and PRs.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by chilkotiKartik** [comment]
### Comprehensive Root Cause Analysis & Two-Layer Resolution

#### Summary of Failure Mechanism
As accurately described in the report, the issue stems from an unhandled **empty constraint intersection** between sequential logit bias processors:
1. `allowed_token_ids` sets all vocabulary logits to `-inf` except the allowed set $S_{	ext{allow}}$.
2. `bad_words` processor subsequently overwrites the exact tokens in $S_{	ext{allow}} \cap S_{	ext{bad}}$ to `-inf`.
3. When $S_{	ext{allow}} \subseteq S_{	ext{bad}}$, the logit vector for the batch row becomes strictly $(-\infty, -\infty, \dots, -\infty)$.
4. `torch.argmax(logits, dim=-1)` under temperature 0 evaluates all ties at index `0`, silently emitting out-of-vocabulary / non-allowed tokens without raising an error.

---

#### Proposed Solution: Two-Layer Defense

##### 1. Upfront Static Request Validation in `vllm/sampling_params.py`
In `SamplingParams.verify_args()`, check for total intersection collision when both `allowed_token_ids` and single-token `bad_words` are specified:

```python
if self.allowed_token_ids is not None and self.bad_words_token_ids is not None:
    allowed_set = set(self.allowed_token_ids)
    # Check if all allowed tokens are prohibited at step 0
    single_token_bad_words = {
        tokens[0] for tokens in self.bad_words_token_ids if len(tokens) == 1
    }
    if allowed_set and allowed_set.issubset(single_token_bad_words):
        raise ValueError(
            f"Contradictory sampling parameters: all allowed_token_ids ({self.allowed_token_ids}) "
            f"are explicitly banned by bad_words. Feasible next-token vocabulary is empty."
        )
```

##### 2. Runtime Sampler All-`-inf` Safety Guard in `vllm/v1/worker/gpu/sample/sampler.py`
For dynamic multi-token bad-words sequences where the intersection only becomes empty at step $t > 0$, guard against all-`-inf` logit rows in the sampling kernel:

```python
def check_feasible_logits(logits: torch.Tensor, request_ids: List[str]):
    # Vectorized max across vocab
    max_logits, _ = torch.max(logits, dim=-1)
    is_infeasible = torch.isneginf(max_logits)
    if is_infeasible.any():
        infeasible_indices = torch.nonzero(is_infeasible).squeeze(-1).tolist()
        infeasible_reqs = [request_ids[i] for i in infeasible_indices]
        raise ValueError(
            f"Sampling constraints resulted in an empty feasible token set (all logits are -inf) "
            f"for requests: {infeasible_reqs}. Terminating request."
        )
```

Adding this check guarantees that vLLM never emits illegal token `0` and provides immediate visibility into contradictory decoding constraints.
